File: src/lib/match.ts
import type { LinearUser, AssigneeMatch } from "./types";

function levenshtein(a: string, b: string): number {
  const m = a.length;
  const n = b.length;
  if (m === 0) return n;
  if (n === 0) return m;
  const dp: number[] = new Array(n + 1);
  for (let j = 0; j <= n; j++) dp[j] = j;
  for (let i = 1; i <= m; i++) {
    let prev = dp[0];
    dp[0] = i;
    for (let j = 1; j <= n; j++) {
      const tmp = dp[j];
      dp[j] = Math.min(
        dp[j] + 1,
        dp[j - 1] + 1,
        prev + (a[i - 1] === b[j - 1] ? 0 : 1)
      );
      prev = tmp;
    }
  }
  return dp[n];
}

function norm(s: string): string {
  return s.toLowerCase().trim().replace(/\s+/g, " ");
}

function similarity(a: string, b: string): number {
  const na = norm(a);
  const nb = norm(b);
  if (!na || !nb) return 0;
  if (na === nb) return 1;
  const maxLen = Math.max(na.length, nb.length);
  return 1 - levenshtein(na, nb) / maxLen;
}

// Score a single hint against a user's name fields. Boost first-name-only matches
// since transcripts usually use first names.
function scoreUser(hint: string, user: LinearUser): number {
  const h = norm(hint);
  const candidates = [
    user.name,
    user.displayName,
    user.email.split("@")[0],
  ].filter(Boolean);

  let best = 0;
  for (const c of candidates) {
    const cn = norm(c);
    if (!cn) continue;
    // exact full match
    best = Math.max(best, similarity(h, cn));
    // first-name match (hint is just "Sarah" → match against "Sarah Chen")
    const firstWord = cn.split(" ")[0];
    if (firstWord) {
      best = Math.max(best, similarity(h, firstWord));
    }
    // substring containment as a floor (e.g. "sarah" in "sarah-c")
    if (cn.includes(h) || h.includes(cn)) {
      best = Math.max(best, 0.85);
    }
  }
  return best;
}

export function matchAssignee(
  hint: string | null,
  users: LinearUser[]
): AssigneeMatch | null {
  if (!hint || users.length === 0) return null;
  let bestUser: LinearUser | null = null;
  let bestScore = 0;
  for (const user of users) {
    const s = scoreUser(hint, user);
    if (s > bestScore) {
      bestScore = s;
      bestUser = user;
    }
  }
  if (!bestUser || bestScore < 0.55) return null;
  return { user: bestUser, score: bestScore };
}

File: app/match.py
"""Fuzzy assignee matching — Python port of `src/lib/match.py` (match.ts).

Levenshtein similarity with a first-name boost, since transcripts usually refer
to people by first name. Behaviour and thresholds are kept identical to the
original TypeScript.
"""

from __future__ import annotations

import re
from typing import Optional

from .models import AssigneeMatch, LinearUser

_WS = re.compile(r"\s+")


def _levenshtein(a: str, b: str) -> int:
    m, n = len(a), len(b)
    if m == 0:
        return n
    if n == 0:
        return m
    dp = list(range(n + 1))
    for i in range(1, m + 1):
        prev = dp[0]
        dp[0] = i
        for j in range(1, n + 1):
            tmp = dp[j]
            dp[j] = min(
                dp[j] + 1,
                dp[j - 1] + 1,
                prev + (0 if a[i - 1] == b[j - 1] else 1),
            )
            prev = tmp
    return dp[n]


def _norm(s: str) -> str:
    return _WS.sub(" ", s.lower().strip())


def _similarity(a: str, b: str) -> float:
    na, nb = _norm(a), _norm(b)
    if not na or not nb:
        return 0.0
    if na == nb:
        return 1.0
    max_len = max(len(na), len(nb))
    return 1 - _levenshtein(na, nb) / max_len


def _score_user(hint: str, user: LinearUser) -> float:
    """Score a hint against a user's name fields, boosting first-name matches."""
    h = _norm(hint)
    candidates = [user.name, user.displayName, user.email.split("@")[0]]

    best = 0.0
    for c in candidates:
        if not c:
            continue
        cn = _norm(c)
        if not cn:
            continue
        # exact full match
        best = max(best, _similarity(h, cn))
        # first-name match (hint "Sarah" -> match against "Sarah Chen")
        first_word = cn.split(" ")[0]
        if first_word:
            best = max(best, _similarity(h, first_word))
        # substring containment as a floor (e.g. "sarah" in "sarah-c")
        if h in cn or cn in h:
            best = max(best, 0.85)
    return best


def match_assignee(hint: Optional[str], users: list[LinearUser]) -> Optional[AssigneeMatch]:
    if not hint or not users:
        return None
    best_user: Optional[LinearUser] = None
    best_score = 0.0
    for user in users:
        s = _score_user(hint, user)
        if s > best_score:
            best_score = s
            best_user = user
    if best_user is None or best_score < 0.55:
        return None
    return AssigneeMatch(user=best_user, score=best_score)
